src/utils/matter-parser.js

```diff
@@ -0,0 +1,98 @@
+const yaml = require('yaml');
+const fsp = require('fs').promises;
+const glob = require('glob');
+const { promisify } = require('util');
+
+/**
+ * Class for parsing a markdown file into frontmatter and content
+ */
+class MatterParser {
+  /**
+   * Creates a markdown file instance
+   * @param {string} fileName the file name
+   * @param {object} metaData the parsed frontmatter
+   * @param {string} content the content body
+   * @param {string} eol the line-ending style used (\n or \r\n)
+   */
+  constructor(fileName, metaData, content, eol = '\n') {
+    this.fileName = fileName;
+    this.metaData = metaData;
+    this.content = content;
+    this.eol = eol;
+  }
+
+  /**
+   * Read file
+   * @param {string} fileName the file name
+   * @returns {Promise<MatterParser>} the parser instance
+   */
+  static async fromFile(fileName) {
+    const content = await fsp.readFile(fileName, 'utf-8');
+    return MatterParser.fromString(content, fileName);
+  }
+
+  /**
+   * Read file from string
+   * @param {string} content contents of the frontmatter file
+   * @returns {MatterParser} the parser instance
+   */
+  static fromString(content = '', fileName = 'output.md') {
+    // determine line endings by looking at the first appearance of \n or \r\n
+    const eol = (content.match(/\n|\r\n/) || ['\n'])[0];
+    // normalize line endings in case of mixed LF/CRLF
+    let normalizedContent = content.replace(/\n|\r\n/g, eol);
+    const marker = '---' + eol;
+    const metaData = {};
+    if (normalizedContent.startsWith(marker)) {
+      const frontMatterEnd = normalizedContent.indexOf(marker, marker.length);
+      const frontMatter = normalizedContent.slice(
+        marker.length,
+        frontMatterEnd
+      );
+      normalizedContent = normalizedContent.slice(
+        frontMatterEnd + marker.length
+      );
+      Object.assign(metaData, yaml.parse(frontMatter));
```

Comment:
Is this worth simplifying with a regular expression maybe?

```js
const expression = (
// Regex fragment        | Meaning
// ======================|========================================
// /                   / | Regular expression delimiters
//  \s*                  | Potential leading spaces/lies
//     ---               | Front-matter start
//        (        )     | Capture group
//         [\s\S]*       | Literally anything any number of times
//                ?      | Greedy, as in as much as we can
//                  ---  | Front-matter end
   /\s*---([\s\S]*?)---/
)
const frontMatter = expression.match(content)
```

Quick demo: https://regexr.com/5ic3j

Reply:
I think it should be a bit more strict:

- the newlines after the 3 hyphens are required
- the front matter should be at the very top of the document (see: https://jekyllrb.com/docs/front-matter/)

`/^---\r?\n([\s\S]*?)\r?\n---\r?\n/`